File: adk/go/weil_go/weil_contracts/fungible/fungible.go
// Package fungible provides a fungible token implementation for Weil contracts.
// It implements a standard ERC-20-like token with transfer, approval, and minting capabilities.
package fungible

import (
	"fmt"

	"github.com/weilliptic-public/wadk/adk/go/weil_go/collections"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/errors"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/ledger"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/runtime"
)

// FungibleToken represents a fungible token contract with name, symbol, total supply, and allowances.
type FungibleToken struct {
	Name        string                              `json:"name"`
	Symbol      string                              `json:"symbol"`
	TotalSupply uint64                              `json:"total_supply"`
	Allowances  collections.WeilMap[string, uint64] `json:"allowances"`
}

// NewFungibleToken creates a new fungible token with the given name and symbol.
// The total supply is initialized to 0.
func NewFungibleToken(name string, symbol string) *FungibleToken {
	return &FungibleToken{
		Name:        name,
		Symbol:      symbol,
		TotalSupply: 0,
		Allowances:  *collections.NewWeilMap[string, uint64](*collections.NewWeilId(0)),
	}
}

// GetName returns the name of the token.
func (t *FungibleToken) GetName() string {
	return t.Name
}

// GetSymbol returns the symbol of the token.
func (t *FungibleToken) GetSymbol() string {
	return t.Symbol
}

// GetTotalSupply returns the total supply of the token.
func (t *FungibleToken) GetTotalSupply() uint64 {
	return t.TotalSupply
}

// BalanceFor retrieves the balance of the token for the given address.
func (t *FungibleToken) BalanceFor(addr string) (*uint64, error) {
	return ledger.BalanceFor(addr, t.Symbol)
}

// Transfer transfers tokens from the sender's address to the specified address.
// Returns an error if the transfer fails (e.g., insufficient balance).
func (t *FungibleToken) Transfer(toAddr string, amount uint64) error {
	sender := runtime.Sender()

	err := ledger.Transfer(t.GetSymbol(), sender, toAddr, amount)
	return err
}

// Approve approves the spender to transfer tokens from the sender's address.
// The approval amount is stored in the allowances map.
func (t *FungibleToken) Approve(spender string, amount uint64) {
	sender := runtime.Sender()
	key := fmt.Sprintf(
		"%s$%s",
		sender,
		spender,
	)
	t.Allowances.Insert(&key, &amount)
}

// Mint creates new tokens and assigns them to the sender's address.
// The total supply is incremented by the minted amount.
// Returns an error if the minting operation fails.
func (t *FungibleToken) Mint(amount uint64) error {
	t.TotalSupply += amount

	sender := runtime.Sender()

	err := ledger.Mint(t.GetSymbol(), sender, amount)
	if err != nil {
		return err
	}

	return nil
}

// TransferFrom transfers tokens from one address to another on behalf of the sender.
// The sender must have been approved to transfer at least the specified amount.
// Returns an error if the transfer fails (e.g., insufficient balance or allowance).
func (t *FungibleToken) TransferFrom(fromAddr string, toAddr string, amount uint64) error {
	spender := runtime.Sender()

	key := fmt.Sprintf(
		"%s$%s",
		fromAddr,
		spender,
	)

	balance, err := t.Allowances.Get(&key)
	if err != nil {
		*balance = 0
	}

	if *balance < amount {
		return errors.NewFunctionReturnedWithError("TransferFrom", nil)
	}

	err = ledger.Transfer(t.GetSymbol(), fromAddr, toAddr, amount)
	if err != nil {
		return err
	}
	remainingBalance := *balance - amount
	t.Allowances.Insert(&key, &remainingBalance)

	return nil
}

// Allowance returns the amount of tokens that the spender is approved to transfer
// from the owner's address. Returns 0 if no allowance has been set.
func (t *FungibleToken) Allowance(owner string, spender string) uint64 {
	key := fmt.Sprintf(
		"%s$%s",
		owner,
		spender,
	)

	balance, err := t.Allowances.Get(&key)
	if err != nil {
		*balance = 0
	}

	return *balance
}

File: adk/asc/weil-sdk/assembly/contracts/fungible.ts
import { Ledger } from '../ledger'
import { WeilMap } from '../collections/map'
import { Runtime } from '../runtime'
import { Result } from '../result'
import { WeilError } from '../error'
import { Box } from '../primitives'
import { WeilId } from '../collections/weil-id'
import { JSON } from 'json-as'
import { Tuple } from '../json/tuples'
import { JSONWrapper } from '../json/primitives'

/**
 * Arguments for the balanceFor method.
 */
@json
export class BalanceForArgs {
  /** The address to query */
  addr!: string
}

/**
 * Arguments for the transfer method.
 */
@json
export class TransferArgs {
  /** The address to transfer to */
  to_addr!: string
  /** The amount to transfer */
  amount!: u64
}

/**
 * Arguments for the approve method.
 */
@json
export class ApproveArgs {
  /** The address to approve as a spender */
  spender!: string
  /** The amount to approve */
  value!: u64
}

/**
 * Arguments for the transferFrom method.
 */
@json
export class TransferFromArgs {
  /** The address to transfer from */
  from_addr!: string
  /** The address to transfer to */
  to_addr!: string
  /** The amount to transfer */
  value!: u64
}

/**
 * Arguments for the allowance method.
 */
@json
export class AllowanceArgs {
  /** The owner address */
  owner!: string
  /** The spender address */
  spender!: string
}

/**
 * Generates a unique key for an allowance mapping.
 * 
 * @param owner - The owner address
 * @param spender - The spender address
 * @returns A unique key string
 */
const getAllowanceKey = (owner: string, spender: string): string =>
  [owner, spender].join('$')

/**
 * FungibleToken class implementing ERC-20-like functionality.
 * Manages token balances, transfers, approvals, and allowances.
 */
@json
export class FungibleToken {
  /** The name of the token */
  name: string
  /** The symbol of the token */
  symbol: string
  /** The total supply of tokens */
  totalSupply: u64
  /** Map of allowance keys to approved amounts */
  allowances: WeilMap<string, Box<u64>>

  /**
   * Creates a new FungibleToken instance.
   * 
   * @param name - The name of the token
   * @param symbol - The symbol of the token
   */
  constructor(name: string, symbol: string) {
    this.name = name
    this.symbol = symbol
    this.totalSupply = 0
    this.allowances = new WeilMap<string, Box<u64>>(new WeilId(0))
  }

  /**
   * Gets the balance of a specific address for this token.
   * 
   * @param addr - The address to query
   * @returns A Result containing the balance or an error
   */
  balanceFor(addr: string): Result<Box<u64>, WeilError> {
    return Ledger.balanceFor(addr, this.symbol)
  }

  /**
   * Transfers tokens from the sender to another address.
   * 
   * @param to_addr - The address to transfer to
   * @param amount - The amount to transfer
   * @returns A Result indicating success or failure
   */
  transfer(to_addr: string, amount: u64): Result<Box<i32>, WeilError> {
    return Ledger.transfer(this.symbol, Runtime.sender(), to_addr, amount)
  }

  /**
   * Approves a spender to transfer tokens on behalf of the sender.
   * 
   * @param spender - The address to approve as a spender
   * @param amount - The amount to approve
   */
  approve(spender: string, amount: u64): void {
    const key = getAllowanceKey(Runtime.sender(), spender)
    this.allowances.set(key, new Box(amount))
  }

  /**
   * Mints new tokens to the sender's address.
   * 
   * @param amount - The amount to mint
   * @returns A Result indicating success or failure
   */
  mint(amount: u64): Result<Box<i32>, WeilError> {
    this.totalSupply+=amount;

    return Ledger.mint(
      this.symbol,
      Runtime.sender(),
      amount
    )
  }

  /**
   * Transfers tokens from one address to another on behalf of the sender.
   * Requires the sender to have sufficient allowance from the from_addr.
   * 
   * @param from_addr - The address to transfer from
   * @param to_addr - The address to transfer to
   * @param amount - The amount to transfer
   * @returns A Result indicating success or failure
   */
  transferFrom(
    from_addr: string,
    to_addr: string,
    amount: u64,
  ): Result<Box<i32>, WeilError> {
    const key = getAllowanceKey(from_addr, Runtime.sender())
    let balance: u64 = 0
    if (this.allowances.has(key)) {
      balance = this.allowances.get(key)!.value
    }

    if (balance < amount) {
      return Result.Err<Box<i32>, WeilError>(
        WeilError.FunctionReturnedWithError({
          method_name: 'transfer_from',
          err_msg: `Allowance balance of sender ${Runtime.sender()} is ${balance}, which is less than the transfer request amount from ${from_addr}.`,
        }),
      )
    }

    const result = Ledger.transfer(this.symbol, from_addr, to_addr, amount)
    this.allowances.set(key, new Box(balance - amount))

    return result
  }

  /**
   * Gets the allowance amount that a spender is approved to spend on behalf of an owner.
   * 
   * @param owner - The owner address
   * @param spender - The spender address
   * @returns The allowance amount, or 0 if no allowance exists
   */
  allowance(owner: string, spender: string): u64 {
    const key = getAllowanceKey(owner, spender)
    return this.allowances.has(key) ? this.allowances.get(key)!.value : 0
  }
}
